R/bootstrap.R

```diff
@@ -949,56 +964,173 @@ buildTabset <- function(tabs, ulClass, textFilter = NULL, id = NULL,
+      } else {
+        # In BS4, dropdown nav anchors can't be wrapped in a <li> tag
+        # and also need .nav-link replaced with .dropdown-item to be
+        # styled sensibly
+        items <- tabset$navList
+        items$children <- lapply(items$children, function(x) {
+          # x should be a tagFunction() due to the else block below
+          x <- if (inherits(x, "shiny.tag.function")) x() else x
+          # Replace <li class="nav-item"><a class="nav-link"></a></li>
+          #    with <a class="dropdown-item"></a>
+          if (tagHasClass(x, "nav-item")) {
+            x <- x$children[[1]]
+            x$attribs$class <- "dropdown-item"
+          }
+          x
+        })
+        bs4NavItemDropdown(title, value, icon, active, items)
+      }
+    })
+  )
+}
 
-  } else {
-    # tabPanel item: create the tab's liTag and divTag
-    tabId <- paste("tab", tabsetId, index, sep = "-")
-    liTag <- tags$li(
-               tags$a(
-                 href = paste("#", tabId, sep = ""),
-                 `data-toggle` = "tab",
-                 `data-value` = divTag$attribs$`data-value`,
-                 getIcon(iconClass = divTag$attribs$`data-icon-class`),
-                 divTag$attribs$title
-               )
+
+bs3NavItemDropdown <- function(title, value, icon, active, items) {
+  tags$li(
+    class = "dropdown",
+    class = if (active) "active", # BS3
+    tags$a(
+      href = "#",
+      class = "dropdown-toggle",
+      `data-toggle` = "dropdown",
+      `data-value` = value,
+      icon,
+      title, tags$b(class = "caret")
+    ),
+    items
+  )
+}
+
+bs3NavItem <- function(id, title, value, icon, active) {
+  tags$li(
+    class = if (active) "active",
+    tags$a(
+      href = paste0("#", id),
+      `data-toggle` = "tab",
+      `data-value` = value,
+      icon,
+      title
     )
-    # if this tabPanel is selected item, mark it active
-    if (isTabSelected(divTag)) {
-      liTag$attribs$class <- "active"
-      divTag$attribs$class <- "tab-pane active"
-    }
-    divTag$attribs$id <- tabId
-    divTag$attribs$title <- NULL
-  }
-  return(list(liTag = liTag, divTag = divTag))
+  )
 }
 
+bs4NavItemDropdown <- function(title, value, icon, active, items) {
+  tags$li(
+    class = "dropdown",
+    class = "nav-item",
+    tags$a(
+      href = "#",
+      class = "dropdown-toggle",
+      class = "nav-link",
+      class = if (active) "active",
+      `data-toggle` = "dropdown",
+      `data-value` = value,
+      icon,
+      title,
+      tags$b(class = "caret") # TODO: can be removed?
+    ),
+    items
+  )
+}
+
+bs4NavItem <- function(id, title, value, icon, active) {
+  tags$li(
+    class = "nav-item",
+    tags$a(
+      class = "nav-link",
+      class = if (active) "active",
+      href = paste0("#", id),
+      `data-toggle` = "tab",
+      `data-value` = value,
+      icon,
+      title
+    )
+  )
+}
+
+# TODO: something like this should exist in htmltools
+tagHasClass <- function(x, class) {
```

Comment:
The complexity of this code inspired this issue:
https://github.com/rstudio/htmltools/issues/205